lib/mumukit/auth/grant.rb

```diff
@@ -29,68 +71,120 @@ def inspect
       "<Mumukit::Auth::Grant #{to_s}>"
     end
 
-    def self.parse(pattern)
-      case pattern
-        when '*' then
-          AllGrant.new
-        when '*/*' then
-          AllGrant.new
-        when /(.*)\/\*/
-          FirstPartGrant.new($1)
-        else
-          SingleGrant.new(Slug.parse pattern)
-      end
+    # Tells wether the given grant
+    # is authorized by this grant
+    #
+    # This method exist in order to implement double dispatching
+    # for both grant and slugs authorization
+    #
+    # See:
+    #  * `Mumukit::Auth::Slug#authorized_by?`
+    #  * `Mumukit::Auth::Grant::Base#allows?
+    #  * `Mumukit::Auth::Grant::Base#includes?`
+    def authorized_by?(grant)
+      grant.includes? self
+    end
+
+    # tells whether the given slug-like object is allowed by
+    # this grant
+    required :allows?
+
+    # tells whether the given grant-like object is included
+    # in - that is, is not broader than - this grant
+    #
+    # :warning: Custom grants **should not** override this method
+    def includes?(grant_like)
+      self == grant_like.to_mumukit_grant
     end
+
+    # Returns a canonical string representation of this grant
+    # Equivalent grant **must** have equivalent string representations
+    required :to_s
   end
 
-  class AllGrant < Grant
-    def allows?(_resource_slug)
+  class AllGrant < Base
+    def allows?(_slug_like)
+      true
+    end
+
+    def includes?(_)
       true
     end
 
     def to_s
       '*'
     end
 
-    def to_mumukit_slug
-      Mumukit::Auth::Slug.new '*', '*'
+    def self.try_parse(pattern)
+      new if ['*', '*/*'].include? pattern
     end
   end
 
-  class FirstPartGrant < Grant
+  class FirstPartGrant < Base
+    attr_accessor :first
+
     def initialize(first)
+      raise Mumukit::Auth::InvalidGrantFormatError, "Invalid first grant. First part must not be _" if  first == '_'
       @first = first.downcase
     end
 
-    def allows?(resource_slug)
-      resource_slug.to_mumukit_slug.normalize!.match_first @first
+    def granted_organizations
+      [first]
+    end
+
+    def allows?(slug_like)
+      slug_like.to_mumukit_slug.normalize!.match_first @first
     end
 
     def to_s
       "#{@first}/*"
     end
 
-    def to_mumukit_slug
-      Mumukit::Auth::Slug.new @first, '*'
+    def includes?(grant_like)
+      grant = grant_like.to_mumukit_grant
+      case grant
+      when FirstPartGrant then grant.first == first
+      when SingleGrant then grant.slug.first == first
```

Comment:
We could delegate `first` to `slug` in `SingleGrant` and compact this two lines like:

`when FirstPartGrant, SingleGrant then grant.first == first`